Project/Gem/Source/VehicleController/VehicleControllerComponent.cpp

```diff
@@ -0,0 +1,326 @@
+        const float crossTrackError = nearestPositionVehicleFrame.Dot(-AZ::Vector3::CreateAxisY());
+        const float lateralError = nearestPositionVehicleFrame.Dot(AZ::Vector3::CreateAxisX()) - directions[m_currentState];
+
+        float steeringAngle = -m_configuration.m_vehicleSteeringGain * crossTrackError;
+        const float speed = m_configuration.m_vehicleSpeed * directions[m_currentState];
+
+        // Reduce steering for large distances (to eliminate "zig-zag" movement)
+        if (lateralError > 1.0f)
+        {
+            steeringAngle /= lateralError;
+        }
+
+        // Apply Ackermann steering
+        ROS2::AckermannCommandStruct acs;
+        acs.m_steeringAngle = steeringAngle;
+        acs.m_steeringAngleVelocity = 0.0f;
+        acs.m_speed = speed;
+        acs.m_acceleration = 0.0f;
+        acs.m_jerk = 0.0f;
+        ROS2::AckermannNotificationBus::Event(m_configuration.m_vehicleEntityId, &ROS2::AckermannNotifications::AckermannReceived, acs);
+
+        return AZStd::make_pair(crossTrackError, lateralError);
+    }
+
+    void VehicleControllerComponent::detectCollisions(const AZ::Vector3& vehiclePosition)
+    {
+        for (const auto& obstacleId : m_configuration.m_predefinedObstacles)
+        {
+            AZ::Vector3 obstacleTranslation = AZ::Vector3::CreateZero();
+            AZ::TransformBus::EventResult(obstacleTranslation, obstacleId, &AZ::TransformBus::Events::GetWorldTranslation);
+
+            if ((obstacleTranslation - vehiclePosition).GetLength() < 5.0)
+            {
+                // Set the flag and stop the vehicle
+                if (!m_obstacleDetected)
+                {
+                    m_obstacleDetected = true;
+                    m_currentState = VehicleState::STOPPED;
+                }
+                // No need to look for other collisions
+                return;
+            }
+        }
+        // No collisions detected; reset the flag
+        m_obstacleDetected = false;
+    }
+
+    void VehicleControllerComponent::switchPath()
+    {
+        if (m_currentPath < m_configuration.m_predefinedPaths.size() - 1)
+        {
+            m_currentState = VehicleState::CHANGING_PATH;
+            ++m_currentPath;
+        }
+        else
+        {
+            m_currentState = VehicleState::FINISHED;
+        }
+    }
+
+    void VehicleControllerComponent::OnTick(float deltaTime, [[maybe_unused]] AZ::ScriptTimePoint timePoint)
+    {
+        if (deltaTime > 0.25f)
+        {
+            return;
+        }
+
+        // startup delay
+        m_currentTime += deltaTime;
+        if (m_currentTime < m_configuration.m_startDelay)
+        {
+            return;
+        }
+
+        // switch on/off lights
+        if (m_switchLights)
+        {
+            lightsOn();
+            m_switchLights = false;
+            m_lightsTime = m_currentTime + 0.2f; // keep the lights on for 200ms
+        }
+        if (m_lightsTime < m_currentTime)
```

Comment:
This if statement is true most of the time. The called function is not computationally heavy but it might be adding some unnecessary overhead. Consider adding a boolean to prevent calling lightsOff on each tick.

Reply:
Setting `m_lightsTime` to infinity should solve the problem.